java.lang.RuntimeException: whoops, better handle OTHER null

Encounter error when use Gradle to compile google android code with Error Prone plugin:
```
src/com/android/providers/settings/SettingsProvider.java:1487: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
return settingsState.new Setting(ssaidSetting) {
^

error-prone version: 2.3.3
BugPattern: NullAway

 Stack Trace:
 java.lang.RuntimeException: whoops, better handle OTHER null
at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:1952)
at com.uber.nullaway.NullAway.handleInvocation(NullAway.java:1352)
at com.uber.nullaway.NullAway.matchNewClass(NullAway.java:354)
at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:433)
at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:759)
at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:150)
at com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1705)
```

Gradle compilation passed without error-prone plugin, but failed if plugin integrated.

Relevant code fragment:

```
private Setting mascaradeSsaidSetting(SettingsState settingsState, Setting ssaidSetting) {
    if (ssaidSetting != null) {
        return settingsState.new Setting(ssaidSetting) {
              @override
               public int getKey() {
                    final int userId = getUserIdFromKey(super.getKey());
                    return makeKey(SETTINGS_TYPE_SECURE, userId);
                  }

             @Override
             public String getName() {
                 return Settings.Secure.ANDROID_ID;
              }
          };
      }
    return null;
}
```

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -1309,6 +1309,13 @@ public class NullAway extends BugChecker
               this, state, methodSymbol, actualParams, ImmutableSet.of());
     }
     List<VarSymbol> formalParams = methodSymbol.getParameters();
+
+    if (formalParams.size() != actualParams.size() && !methodSymbol.isVarArgs()) {
+      // In some special cases like one in issue #366
+      // formal params and actual params do not match while using JDK11
+      return Description.NO_MATCH;
+    }
+
     if (nonNullPositions == null) {
       ImmutableSet.Builder<Integer> builder = ImmutableSet.builder();
       // compute which arguments are @NonNull
@@ -1330,6 +1337,7 @@ public class NullAway extends BugChecker
       }
       nonNullPositions = builder.build();
     }
+
     // now actually check the arguments
     // NOTE: the case of an invocation on a possibly-null reference
     // is handled by matchMemberSelect()

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -25,6 +25,7 @@ package com.uber.nullaway;
 import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
 import static com.sun.source.tree.Tree.Kind.EXPRESSION_STATEMENT;
 import static com.sun.source.tree.Tree.Kind.IDENTIFIER;
+import static com.sun.source.tree.Tree.Kind.OTHER;
 import static com.sun.source.tree.Tree.Kind.PARENTHESIZED;
 import static com.sun.source.tree.Tree.Kind.TYPE_CAST;
 import static com.uber.nullaway.ErrorBuilder.errMsgForInitializer;
@@ -2095,7 +2096,7 @@ public class NullAway extends BugChecker
   }
 
   /**
-   * strip out enclosing parentheses and type casts.
+   * strip out enclosing parentheses, type casts and Nullchk operators.
    *
    * @param expr
    * @return
@@ -2110,6 +2111,12 @@ public class NullAway extends BugChecker
       }
       if (expr.getKind().equals(TYPE_CAST)) {
         expr = ((TypeCastTree) expr).getExpression();
+        someChange = true;
+      }
+
+      // Strips Nullchk operator
+      if (expr.getKind().equals(OTHER) && expr instanceof JCTree.JCUnary) {
+        expr = ((JCTree.JCUnary) expr).getExpression();
         someChange = true;
       }
     }

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -1310,9 +1310,14 @@ public class NullAway extends BugChecker
     }
     List<VarSymbol> formalParams = methodSymbol.getParameters();
 
-    if (formalParams.size() != actualParams.size() && !methodSymbol.isVarArgs()) {
-      // In some special cases like one in issue #366
-      // formal params and actual params do not match while using JDK11
+    if (formalParams.size() != actualParams.size()
+        && !methodSymbol.isVarArgs()
+        && !methodSymbol.isStatic()
+        && methodSymbol.isConstructor()
+        && methodSymbol.enclClass().isInner()) {
+      // In special cases like one in issue #366
+      // formal params and actual params do not match while using JDK11+
+      // we bail out in this particular case
       return Description.NO_MATCH;
     }
 

--- a/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java
+++ b/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java
@@ -378,7 +378,10 @@ public class AccessPathNullnessPropagation implements TransferFunction<Nullness,
   @Override
   public TransferResult<Nullness, NullnessStore> visitNullChk(
       NullChkNode nullChkNode, TransferInput<Nullness, NullnessStore> input) {
-    throw new RuntimeException("we should never see this");
+    SubNodeValues values = values(input);
+    Nullness nullness =
+        hasPrimitiveType(nullChkNode) ? NONNULL : values.valueOfSubNode(nullChkNode.getOperand());
+    return noStoreChanges(nullness, input);
   }
 
   @Override

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
@@ -918,4 +918,44 @@ public class NullAwayNegativeCases {
     s += boxAndDeref(n >>> 3);
     return s;
   }
+
+  // Cases from issue https://github.com/uber/NullAway/issues/366
+  class AnotherClass {
+    class Inner {
+      int x;
+    }
+  }
+
+  AnotherClass.Inner anotherMethod() {
+    AnotherClass anotherClass = new AnotherClass();
+    return anotherClass.new Inner() {
+      int getX() {
+        return x;
+      }
+    };
+  }
+
+  class YetAnotherClass {
+    class Inner {
+      int x;
+
+      Inner(int __x) {
+        x = __x;
+      }
+
+      int getX() {
+        return 3;
+      }
+    }
+  }
+
+  YetAnotherClass.Inner yetAnotherMethod() {
+    YetAnotherClass yetAnotherClass = new YetAnotherClass();
+    return yetAnotherClass.new Inner(4) {
+      @Override
+      int getX() {
+        return 5;
+      }
+    };
+  }
 }

